File: frontend/src/types/hospital.ts
export type OperationalPriority = 'Critical' | 'High' | 'Moderate' | 'Low';

export type CaseStatus = 
  | 'Alerted'
  | 'Acknowledged'
  | 'Preparing'
  | 'En Route'
  | 'Arrived'
  | 'Handover'
  | 'Diverted'
  | 'Closed';

export type EmsTransportStatus = 
  | 'Dispatched'
  | 'En Route to Scene'
  | 'On Scene'
  | 'Transporting'
  | 'Arrived at Bay'
  | 'Handover Complete';

export type ConsciousnessLevel = 'Alert' | 'Voice' | 'Pain' | 'Unresponsive';

export type ResourceStatus = 'Ready' | 'Limited' | 'Unavailable' | 'Occupied';

export type ResourceCategory = 
  | 'Emergency Bay'
  | 'Resuscitation Unit'
  | 'Specialist Team'
  | 'Critical Equipment'
  | 'Diagnostics / Imaging'
  | 'Surgical / OR'
  | 'Blood Bank';

export interface PreArrivalVitals {
  heartRate?: number;
  systolicBp?: number;
  diastolicBp?: number;
  respiratoryRate?: number;
  oxygenSaturation?: number;
  gcs?: number;
  temperature?: number;
  painScale?: number; // 1-10
  lastUpdated?: string;
  source: 'Reported / Bystander' | 'EMS Verified';
}

export interface PreArrivalClinicalSummary {
  chiefComplaint: string;
  reportedSymptoms: string[];
  pertinentNegatives: string[];
  mechanismOfInjury?: string;
  symptomOnsetMinutes?: number;
  consciousness: ConsciousnessLevel;
  knownAllergies: string[];
  knownMedications: string[];
  relevantHistory: string[];
  initialObservations: string;
  aiStructuringNotes?: string;
  specialtyRequirements: string[];
}

export interface EmsUnitInfo {
  unitId: string;
  vehicleType: 'ALS (Advanced Life Support)' | 'BLS (Basic Life Support)' | 'MICU (Mobile Intensive Care)';
  crewNames: string[];
  crewContact: string;
  currentLocationName: string;
  speedKmH: number;
  distanceRemainingKm: number;
  etaMinutes: number;
  transportStatus: EmsTransportStatus;
  lastTelemetryTimestamp: string;
  verifiedVitals?: PreArrivalVitals;
}

export interface TimelineEvent {
  id: string;
  timestamp: string;
  relativeTime: string;
  title: string;
  description: string;
  actor: string;
  actorRole: string;
  stage: CaseStatus;
  type: 'system' | 'ems' | 'hospital' | 'bystander';
}

export interface EmergencyCase {
  id: string; // e.g., 'LS-2026-001'
  incidentType: string;
  operationalPriority: OperationalPriority;
  status: CaseStatus;
  patientAge: number;
  patientSex: 'Male' | 'Female' | 'Unknown';
  patientCount: number;
  reportedLocation: string;
  destinationHospitalId: string;
  destinationHospitalName: string;
  assignedBay?: string;
  timeReported: string;
  timeAlerted: string;
  timeAcknowledged?: string;
  acknowledgedBy?: string;
  divertedAt?: string;
  divertReason?: string;
  divertNotes?: string;
  clinicalSummary: PreArrivalClinicalSummary;
  emsUnit: EmsUnitInfo;
  timeline: TimelineEvent[];
  readinessChecklist: {
    id: string;
    label: string;
    completed: boolean;
    requiredFor: string;
  }[];
}

export interface HospitalResourceItem {
  id: string;
  name: string;
  category: ResourceCategory;
  status: ResourceStatus;
  totalCapacity: number;
  availableCapacity: number;
  assignedCaseId?: string;
  location: string;
  lastUpdated: string;
  notes?: string;
}

export interface HospitalProfile {
  id: string;
  name: string;
  shortName: string;
  traumaLevel: string;
  operationalStatus: 'Operational' | 'Degraded' | 'Offline';
  diversionActive: boolean;
  activeSurgeLevel: 'Normal' | 'Level 1 Surge' | 'Level 2 Surge' | 'Maximum Capacity';
  totalBays: number;
  availableBays: number;
  onDutyCoordinator: string;
  coordinatorRole: string;
  phone: string;
  address: string;
  capabilities: {
    name: string;
    available: boolean;
    description: string;
  }[];
}

File: backend/app/models/hospital.py
from sqlalchemy import Column, String, Integer, Float, Boolean, ForeignKey, Text, JSON
from sqlalchemy.orm import relationship
from app.core.database import Base
from app.models.base import TimestampMixin


class Hospital(Base, TimestampMixin):
    """
    Receiving hospital facility profile with capabilities and capacity attributes.
    """
    __tablename__ = "hospitals"

    id = Column(String(64), primary_key=True)  # e.g. HOSP-CITYCARE-01
    name = Column(String(255), nullable=False)
    short_name = Column(String(64), nullable=False)
    trauma_level = Column(String(128), nullable=False)

    operational_status = Column(String(32), default="Operational", nullable=False)
    diversion_active = Column(Boolean, default=False, nullable=False)
    active_surge_level = Column(String(64), default="Normal", nullable=False)

    total_bays = Column(Integer, default=12, nullable=False)
    available_bays = Column(Integer, default=5, nullable=False)

    coordinator_name = Column(String(128), nullable=False)
    coordinator_role = Column(String(128), nullable=False)
    phone = Column(String(64), nullable=False)
    address = Column(String(255), nullable=False)
    latitude = Column(Float, nullable=True)
    longitude = Column(Float, nullable=True)

    capabilities = Column(JSON, default=list, nullable=False)

    # Relationships
    resources = relationship("HospitalResource", back_populates="hospital", cascade="all, delete-orphan")
    assigned_cases = relationship("EmergencyCase", back_populates="destination_hospital")


class HospitalResource(Base, TimestampMixin):
    """
    Specific hospital readiness resource (Resuscitation bay, trauma bay, cath lab, CT scanner, blood bank).
    """
    __tablename__ = "hospital_resources"

    id = Column(String(64), primary_key=True)  # e.g. RES-BAY-01
    hospital_id = Column(String(64), ForeignKey("hospitals.id", ondelete="CASCADE"), nullable=False, index=True)

    name = Column(String(128), nullable=False)
    category = Column(String(64), nullable=False)  # Resuscitation Unit, Emergency Bay, Specialist Team, Diagnostics / Imaging, Blood Bank
    status = Column(String(32), default="Ready", nullable=False)  # Ready, Limited, Occupied, Unavailable

    total_capacity = Column(Integer, default=1, nullable=False)
    available_capacity = Column(Integer, default=1, nullable=False)
    assigned_case_id = Column(String(32), nullable=True)
    location = Column(String(128), nullable=False)
    notes = Column(Text, nullable=True)
    last_updated = Column(String(64), default="Just now", nullable=False)

    # Relationships
    hospital = relationship("Hospital", back_populates="resources")
